File: output/csbc_semish_2026/mistralai_mistral_small_2603/few/cpp/Capitais.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> adj;
vector<int> degree;
vector<int> leaves;

void find_leaves(int n) {
    for (int i = 1; i <= n; i++) {
        if (degree[i] == 1) {
            leaves.push_back(i);
        }
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n;
    cin >> n;

    adj.resize(n + 1);
    degree.resize(n + 1, 0);

    for (int i = 0; i < n - 1; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
        degree[u]++;
        degree[v]++;
    }

    find_leaves(n);

    int max_dist = 0;
    queue<pair<int, int>> q;

    for (int leaf : leaves) {
        q.push({leaf, 0});
    }

    vector<bool> visited(n + 1, false);
    while (!q.empty()) {
        auto [node, dist] = q.front();
        q.pop();

        if (visited[node]) continue;
        visited[node] = true;

        if (degree[node] == 1 && node != leaf) {
            max_dist = max(max_dist, dist);
        }

        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                q.push({neighbor, dist + 1});
            }
        }
    }

    cout << max_dist << endl;

    return 0;
}

File: output/csbc_semish_2026/mistralai_mistral_small_2603/few/python/Capitais.py
import sys
from collections import deque

def main():
    data = sys.stdin.read().split()
    if not data:
        return

    n = int(data[0])
    adj = [[] for _ in range(n + 1)]
    idx = 1
    for _ in range(n - 1):
        u = int(data[idx])
        v = int(data[idx + 1])
        adj[u].append(v)
        adj[v].append(u)
        idx += 2

    leaves = []
    degree = [0] * (n + 1)
    for i in range(1, n + 1):
        degree[i] = len(adj[i])
        if degree[i] == 1:
            leaves.append(i)

    capitals = []
    for i in range(1, n + 1):
        if degree[i] == 1:
            capitals.append(i)

    if len(capitals) == 2:
        print(1)
        return

    q = deque()
    dist = [-1] * (n + 1)
    for leaf in leaves:
        dist[leaf] = 0
        q.append(leaf)

    while q:
        u = q.popleft()
        for v in adj[u]:
            if dist[v] == -1:
                dist[v] = dist[u] + 1
                q.append(v)

    max_dist = 0
    for capital in capitals:
        if dist[capital] > max_dist:
            max_dist = dist[capital]

    print(max_dist)

if __name__ == "__main__":
    main()